### ISSUE TITLE
Details missing in the paper?

### ISSUE THREAD

**Issue by seann999** [chenxy99/Generative-Temporal-Models-with-Spatial-Memory#3]
I did not read anything about a matrix loss in the paper.
Can you explain what it does or point to any resource?
Thanks

Edit: There also seems to be training phases in the train.py that I did not see mentioned in the paper...

**Comment by chenxy99** [comment]
Actually, we can eliminate the use of matrix loss, and we can still get the corresponding result.
![img](https://user-images.githubusercontent.com/23704081/58365700-a368f800-7efa-11e9-95dd-a3565e941fd3.jpg)
We find that in the inferred states, if we do not use the matrix loss, the up/down vectors and left/right vectors are not mutually perpendicular, and the amplitude of up/down vectors is not equal to that of the left/right vectors which conflicts with our priori knowledge.

Furthermore, I used to ask the author Marco Fraccaro about rationality of the matrix loss. 
Very thanks for the author's enthusiastic help. Here is part of his response.
> I do not force orthogonality and the norm constraint in any of the experiments. In other runs I was getting similar plots as in your first figure! This is still correct, as we are in a latent space and we have no control on how the latent manifold will be: it can be rotated/stretched/translated with respect to the original one.
Adding penalty terms as you did is however also good as they basically force the space to be better suited for the euclidean distance used in the DND (unless you learn the metric to be used in the nearest neighbor search).

**Comment by seann999** [comment]
Oh ok, so the matrix loss and the training phases based on epochs are there mostly to make the state space "look good". Thanks!

**Comment by seann999** [comment]
I have another question about something in the code not mentioned in the paper.
The code has 2 ways to sample trajectories: "without wall" changes actions to "no action" when bumping into a wall and "with wall" does not. The code starts training without walls and disables learning magnitude of state changes (sigmoid), and then after some epochs (like epoch 15) it switches the walls on and enables sigmoid training.
I assume this makes training easier, but was it necessary?
Can the model be trained with "with wall" and sigmoid training only? I can experiment myself, but I'd like to know if you have tried it, especially since the paper does not mention this technique.

**Comment by chenxy99** [comment]
Yes. At first, I follow the setting mentioned by the paper, eg. direcdtly training the model with wall and sigmoid simultaneously. But unfortunately, I cannot get the reasonable results, since the sigmoid can hardly match the existance of the wall, which costs me a long time to tune .

Therefore, I emailed the author for help and he replied.

> As I am sure you are aware the implementation can be tricky. To make sure everything worked as expected I found it very useful to start from very simple experiments and make them progressively harder, I recommend you to do the same.
The simplest experiment you can try to see if you can learn the transitions is to only use linear transitions. To do this you need to make sure that there are no "walls" in your data generator, i.e. that if the agent is touching the right border of the image it will not be allowed to go right in the next iteration. In this way you do not need the sigma_d network and can only try to learn M.
Once this works, you can then add walls to the environment and try to learn the neural network as well, first using the action displacements and then with one-hot actions.

Then, based on the help of the author, I can get the similar result as the paper demonstrates.

**Comment by seann999** [comment]
Ok, I tried it with the paper settings in the code and got the results you described; it seems like a crucial detail (maybe the paper should have mentioned it 😐)
Thanks!
